### PR TITLE
`x-nmos-api` claim validation

### PR DESCRIPTION
**Pull request by dannymeloy**
- Add x-nmos-api claims validation function
- Add ability to pull JWKS endpoint from server metadata if available
- Alter claims options for each API
- Tidy up authlib errorhandler

### CODE REVIEW

File: nmoscommon/auth/claims_options.py [hunk 1]
```diff
@@ -1,90 +1,74 @@
...
+    "nbf": {"essential": False},
+    "client_id": {"essential": True},
+    "scope": {"essential": False},
+    "x-nmos-api": {
+        "essential": True,
+        "values": ["registration", "query", "node", "connection"]
+    }
 }
 
 IS_04_REG_CLAIMS = {
```
Comment (andrewbonney) [inline nmoscommon/auth/claims_options.py]: How do you envisage these being used? Is there no way to reduce the duplication a little, perhaps by copying the IS_XX_CLAIMS object and overriding part of it?

File: nmoscommon/auth/claims_validator.py [hunk 2]
```diff
@@ -32,19 +35,51 @@ def validate_aud(self):  # placeholder
...
-            if actual_claim_value.get(attr) != valid_claim_value.get(attr):
-                raise InvalidClaimError(claim_name)
+        def _validate_permissions_object(valid_api_value):
+            access_permission_object = actual_claim_value.get(valid_api_value)
+            if not access_permission_object:
+                raise InvalidClaimError("{}. No entry in claim for '{}'.".format(claim_name, valid_api_value))
+
+            if request.method in ["GET", "OPTIONS", "HEAD"]:
+                access_right = "read"
+            else:
```
Comment (andrewbonney) [inline nmoscommon/auth/claims_validator.py]: Might be worth being explicit here about the write methods? If we get an unexpected method we don't want to give any implicit permission.

File: nmoscommon/auth/claims_validator.py [hunk 3]
```diff
@@ -32,19 +35,51 @@ def validate_aud(self):  # placeholder
...
+                access_right = "write"
+            url_access_list = access_permission_object.get(access_right)
+            if not url_access_list:
+                raise InvalidClaimError(
+                    "{}. No entry in permissions object for '{}'.".format(claim_name, access_right))
+
+            pattern = compile(r'/x-nmos/[a-zA-Z]+/v[0-9]\.[0-9]/(.*)')  # Capture path after namespace
```
Comment (andrewbonney) [inline nmoscommon/auth/claims_validator.py]:
```suggestion
            pattern = compile(r'/x-nmos/[a-z]+/v[0-9]+\.[0-9]+/(.*)')  # Capture path after namespace
```

File: nmoscommon/auth/claims_validator.py [hunk 4]
```diff
@@ -32,19 +35,51 @@ def validate_aud(self):  # placeholder
...
+            trimmed_path = pattern.match(request.path).group(1)
+            for wildcard_url in url_access_list:
+                if fnmatch(trimmed_path, wildcard_url):
+                    return True
+
+            raise InvalidClaimError("{}. No matching paths in access list for '{}'.".format(claim_name, trimmed_path))
+
+        # Single value present in claims options
+        valid_api_value = valid_claim_option.get('value')
```
Comment (andrewbonney) [inline nmoscommon/auth/claims_validator.py]: It's not obvious to me what this is doing. Could you add further comments by way of example, or change the variable names if appropriate?

File: nmoscommon/auth/claims_validator.py [hunk 5]
```diff
@@ -32,19 +35,51 @@ def validate_aud(self):  # placeholder
...
+            _validate_permissions_object(valid_api_value)
+
+        # Multiple existing values present in claims options
+        valid_api_values = valid_claim_option.get('values')
+        if valid_api_values:
+            if not any(api_name in actual_claim_value.keys() for api_name in valid_api_values):
+                raise InvalidClaimError("{}. No entry in claim for one of '{}'.".format(claim_name, valid_api_value))
+            shared_keys = list(set(actual_claim_value.keys()) & set(valid_api_values))  # Find shared keys
+            for valid_api_value in shared_keys:
+                _validate_permissions_object(valid_api_value)
```
Comment (andrewbonney) [inline nmoscommon/auth/claims_validator.py]: Is the 'aud' claim checked anywhere here?

File: nmoscommon/auth/nmos_auth.py [hunk 6]
```diff
@@ -60,20 +58,40 @@ def __init__(self, condition=OAUTH_MODE, claimsOptions=IS_XX_CLAIMS):
         self.claimsOptions = claimsOptions
```
Comment (andrewbonney) [inline nmoscommon/auth/nmos_auth.py]: I can't comment on the specific line in this PR, but just thinking about it, do we end up with an mDNS bridge client instance for every API path which uses the RequiresAuth decorator? If so is there any way we can share a single instance between them? Similar for the rest of the class, does this result in requests for server metadata and keys each time an API path is hit, or is it cached and shared? If cached, is there a procedure for refreshing the keys when they approach expiry?

File: nmoscommon/auth/claims_validator.py (hunk 5 above)
Comment (dannymeloy) [reply nmoscommon/auth/claims_validator.py]: No I've just been working on the x-nmos-api claim for now but this is next on the list

File: nmoscommon/auth/nmos_auth.py (hunk 6 above)
Comment (dannymeloy) [reply nmoscommon/auth/nmos_auth.py]: Sharing a single instance of the mDDNSClient should be fine. WRT the keys, the public key is cached _per instance_ of RequiresAuth (so for each endpoint) but not for all instances within a given API. Might this be resolved with global vars? Might this mean that _every_ API running on a given device might then be using the same keys? That may well not be a problem

Comment (andrewbonney) [reply nmoscommon/auth/nmos_auth.py]: I don't think that is a problem. Using multiple auth servers within a single NMOS deployment isn't specified so there's no need for us to handle it here.

**Pull request merged**